### ISSUE TITLE
[prefer-readonly-parameter-types] Feature request: ignore arrow functions

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3615]
**Repro**

```TS
[].every(([x, y]) => x === y);
```

**Expected Result**

I suggest to add an option to ignore arrow functions, like it is done in "typedef" rule.

Motivation:
1. Usually arrow functions are not reused in many places. On the contrary, they are used locally. So, some people (including me) may think that the downside of having longer code outweights the benefits offered by these rule and may want to turn it off for arrow functions.
2. "typedef" allows to exclude arrow functions. Why? The same reasons apply to prefer-readonly-parameter-types rule.

**Actual Result**

There is no way to exclude arrow functions from this rule. So, one would need to correct the above code to:
```TS
[].every(([x, y]: readonly [unknown, unknown]) => x === y);
```
I don't think this adds any safety. But it makes code less readable.

**Comment by bradzacher** [comment]
> Usually arrow functions are not reused in many places

In your codebase, maybe. Many people use them all over the place in place of function declarations.

> "typedef" allows to exclude arrow functions. Why?

Typedef is a rule which should not be used in any sophisticated codebase. It is designed to help people transition from pure JS to TS. After that it should be removed from the codebase.

TBH I would delete it if it didn't have that use. It's a bad rule and shouldn't be compared against.

---

Imagine the following code instead:

```ts
declare const foo: Array<[Set<string>, Array<string>]>;

foo.every(([x, y]) => {
  x.add('a');
  y.push('b');
  return x.has('c');
})
```

Now all of a sudden you have an arrow function which is "only used locally" but also has very real side-effects.

---

**Comment by ilyub** [comment]
Thx for feedback.

> Now all of a sudden you have an arrow function which is "only used locally" but also has very real side-effects.

I think there is big difference between:
1. Local function effecting ONLY outer scope. In this case you view this function as a part of outer scope.
2.  Global function that is used in remote places, so you don't see function body.
So, I still think function body.

Also, I would probably add an option to exclude types by RegExp. There are types that are two complex to make them readonly, even using DeepReadonly. For example - Vue.

Also, I found that the rule misses on types that use " | " operator.

**Comment by bradzacher** [comment]
> Also, I would probably add an option to exclude types by RegExp.

There's a reason almost none of our rules have one of these.
Options that allowlist/blocklist types by name are a hack at best. They lack context and can easily lead you to accidentally introducing false-NEGATIVES into your codebase.
False-negatives are insidious because they are completely invisible until you find them.

---


Another example of what you will ignore with that sort of option:

```ts
class Foo {
  // oops this is an arrow function that's now ignored
  autoBoundMethod = (arg: Array<string>) => {
    arg.push('hello');
  };
}
```

We won't be adding this option as ignoring all arrow functions is a **huge**, widely impacting change which can and will introduce many, many false negatives.
